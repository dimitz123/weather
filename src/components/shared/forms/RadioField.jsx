import React from 'react';
import PropTypes from 'prop-types';

const RadioOption = ({
  name, groupName, value, active, onChange,
}) => {
  const radioLabel = `${groupName}_${value}`;
  return (
    <label htmlFor={radioLabel} className="radio-option">
      {name}
      <input
        id={radioLabel}
        type="radio"
        name={groupName}
        checked={active}
        onChange={() => onChange(value)}
      />
      <span className="checkmark" />
    </label>
  );
};

RadioOption.propTypes = {
  name: PropTypes.string.isRequired,
  groupName: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.bool,
  ]).isRequired,
  active: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
};

const RadioField = ({
  name, title, value, options, onChange,
}) => {
  const handleChange = (val) => onChange(name, val);
  const formId = `${name}_radio_field`;

  return (
    <div className="form-field">
      <label htmlFor={formId}>
        <span className="label-text">{title}</span>
        <div className="radio-options">
          {options.map((opt) => (
            <RadioOption
              key={opt.value}
              name={opt.name}
              groupName={name}
              value={opt.value}
              active={value === opt.value}
              onChange={handleChange}
            />
          ))}
        </div>
      </label>
    </div>
  );
};

RadioField.propTypes = {
  name: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.bool,
  ]).isRequired,
  options: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.bool,
    ]).isRequired,
  })).isRequired,
  onChange: PropTypes.func.isRequired,
};

export default RadioField;
