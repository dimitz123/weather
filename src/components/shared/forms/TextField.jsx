import React from 'react';
import PropTypes from 'prop-types';

const TextField = ({
  name, title, value, onChange,
}) => {
  const handleChange = (e) => onChange(name, e.target.value);
  const formId = `${name}_text_field`;

  return (
    <div className="form-field">
      <label htmlFor={formId}>
        <span className="label-text">{title}</span>
        <input id={formId} type="text" value={value} name={name} onChange={handleChange} />
      </label>
    </div>
  );
};

TextField.propTypes = {
  name: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
};

export default TextField;
