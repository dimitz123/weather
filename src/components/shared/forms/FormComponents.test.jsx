/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme, { mount } from 'enzyme';
import TextField from './TextField';
import RadioField from './RadioField';

Enzyme.configure({ adapter: new Adapter() });

const textProps = {
  title: 'Test Text',
  name: 'test_text',
  value: 'Testing text field',
  onChange: jest.fn(),
};

it('correctly renders initial TextField properties', () => {
  const textfield = renderer.create(
    <TextField {...textProps} />,
  );

  const tree = textfield.toJSON();
  expect(tree).toMatchSnapshot();
});

it('correctly fires its callback on TextField change', () => {
  const textfield = mount(<TextField {...textProps} />);
  textfield.find('input').simulate('change', { target: { value: 'New Text' } });

  expect(textProps.onChange).toHaveBeenCalledWith('test_text', 'New Text');
});

const radioProps = {
  title: 'Radio Test',
  name: 'test_radio',
  value: 'op1',
  options: [
    { name: 'Op1', value: 'op1' },
    { name: 'Op2', value: 'op2' },
    { name: 'Op3', value: 'op3' },
    { name: 'Op4', value: 'op4' },
  ],
  onChange: jest.fn(),
};

it('renders the RadioField correctly based on the selected value', () => {
  let radiofield = renderer.create(
    <RadioField {...radioProps} />,
  );

  let tree = radiofield.toJSON();
  expect(tree).toMatchSnapshot();

  // Change the selected value and make sure the component correctly re-renders
  radiofield = renderer.create(
    <RadioField {...radioProps} value="op3" />,
  );
  tree = radiofield.toJSON();
  expect(tree).toMatchSnapshot();
});

it('correctly fires RadioField callbacks on change', () => {
  const radiofield = mount(<RadioField {...radioProps} />);

  // Select the current option and make sure no change event is fired
  radiofield.find('label[htmlFor="test_radio_op1"] input').simulate('change');
  expect(radioProps.onChange).toHaveBeenCalledWith('test_radio', 'op1');
  expect(radiofield).toMatchSnapshot();

  radiofield.find('label[htmlFor="test_radio_op3"] input').simulate('change');
  expect(radioProps.onChange).toHaveBeenCalledWith('test_radio', 'op3');
  expect(radiofield).toMatchSnapshot();
});
