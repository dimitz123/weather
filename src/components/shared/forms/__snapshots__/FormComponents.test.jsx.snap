// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioField correctly fires its callback on change 1`] = `
<RadioField
  name="test_radio"
  onChange={
    [MockFunction] {
      "calls": Array [
        Array [
          "test_radio",
          "op1",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  options={
    Array [
      Object {
        "name": "Op1",
        "value": "op1",
      },
      Object {
        "name": "Op2",
        "value": "op2",
      },
      Object {
        "name": "Op3",
        "value": "op3",
      },
      Object {
        "name": "Op4",
        "value": "op4",
      },
    ]
  }
  title="Radio Test"
  value="op1"
>
  <div
    className="form-field"
  >
    <label
      htmlFor="test_radio_radio_field"
    >
      <span
        className="label-text"
      >
        Radio Test
      </span>
      <div
        className="radio-options"
      >
        <RadioOption
          active={true}
          groupName="test_radio"
          key="op1"
          name="Op1"
          onChange={[Function]}
          value="op1"
        >
          <label
            className="radio-option"
            htmlFor="test_radio_op1"
          >
            Op1
            <input
              checked={true}
              id="test_radio_op1"
              name="test_radio"
              onChange={[Function]}
              type="radio"
            />
            <span
              className="checkmark"
            />
          </label>
        </RadioOption>
        <RadioOption
          active={false}
          groupName="test_radio"
          key="op2"
          name="Op2"
          onChange={[Function]}
          value="op2"
        >
          <label
            className="radio-option"
            htmlFor="test_radio_op2"
          >
            Op2
            <input
              checked={false}
              id="test_radio_op2"
              name="test_radio"
              onChange={[Function]}
              type="radio"
            />
            <span
              className="checkmark"
            />
          </label>
        </RadioOption>
        <RadioOption
          active={false}
          groupName="test_radio"
          key="op3"
          name="Op3"
          onChange={[Function]}
          value="op3"
        >
          <label
            className="radio-option"
            htmlFor="test_radio_op3"
          >
            Op3
            <input
              checked={false}
              id="test_radio_op3"
              name="test_radio"
              onChange={[Function]}
              type="radio"
            />
            <span
              className="checkmark"
            />
          </label>
        </RadioOption>
        <RadioOption
          active={false}
          groupName="test_radio"
          key="op4"
          name="Op4"
          onChange={[Function]}
          value="op4"
        >
          <label
            className="radio-option"
            htmlFor="test_radio_op4"
          >
            Op4
            <input
              checked={false}
              id="test_radio_op4"
              name="test_radio"
              onChange={[Function]}
              type="radio"
            />
            <span
              className="checkmark"
            />
          </label>
        </RadioOption>
      </div>
    </label>
  </div>
</RadioField>
`;

exports[`RadioField correctly fires its callback on change 2`] = `
<RadioField
  name="test_radio"
  onChange={
    [MockFunction] {
      "calls": Array [
        Array [
          "test_radio",
          "op1",
        ],
        Array [
          "test_radio",
          "op3",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  options={
    Array [
      Object {
        "name": "Op1",
        "value": "op1",
      },
      Object {
        "name": "Op2",
        "value": "op2",
      },
      Object {
        "name": "Op3",
        "value": "op3",
      },
      Object {
        "name": "Op4",
        "value": "op4",
      },
    ]
  }
  title="Radio Test"
  value="op1"
>
  <div
    className="form-field"
  >
    <label
      htmlFor="test_radio_radio_field"
    >
      <span
        className="label-text"
      >
        Radio Test
      </span>
      <div
        className="radio-options"
      >
        <RadioOption
          active={true}
          groupName="test_radio"
          key="op1"
          name="Op1"
          onChange={[Function]}
          value="op1"
        >
          <label
            className="radio-option"
            htmlFor="test_radio_op1"
          >
            Op1
            <input
              checked={true}
              id="test_radio_op1"
              name="test_radio"
              onChange={[Function]}
              type="radio"
            />
            <span
              className="checkmark"
            />
          </label>
        </RadioOption>
        <RadioOption
          active={false}
          groupName="test_radio"
          key="op2"
          name="Op2"
          onChange={[Function]}
          value="op2"
        >
          <label
            className="radio-option"
            htmlFor="test_radio_op2"
          >
            Op2
            <input
              checked={false}
              id="test_radio_op2"
              name="test_radio"
              onChange={[Function]}
              type="radio"
            />
            <span
              className="checkmark"
            />
          </label>
        </RadioOption>
        <RadioOption
          active={false}
          groupName="test_radio"
          key="op3"
          name="Op3"
          onChange={[Function]}
          value="op3"
        >
          <label
            className="radio-option"
            htmlFor="test_radio_op3"
          >
            Op3
            <input
              checked={false}
              id="test_radio_op3"
              name="test_radio"
              onChange={[Function]}
              type="radio"
            />
            <span
              className="checkmark"
            />
          </label>
        </RadioOption>
        <RadioOption
          active={false}
          groupName="test_radio"
          key="op4"
          name="Op4"
          onChange={[Function]}
          value="op4"
        >
          <label
            className="radio-option"
            htmlFor="test_radio_op4"
          >
            Op4
            <input
              checked={false}
              id="test_radio_op4"
              name="test_radio"
              onChange={[Function]}
              type="radio"
            />
            <span
              className="checkmark"
            />
          </label>
        </RadioOption>
      </div>
    </label>
  </div>
</RadioField>
`;

exports[`RadioField renders correctly based on the selected value 1`] = `
<div
  className="form-field"
>
  <label
    htmlFor="test_radio_radio_field"
  >
    <span
      className="label-text"
    >
      Radio Test
    </span>
    <div
      className="radio-options"
    >
      <label
        className="radio-option"
        htmlFor="test_radio_op1"
      >
        Op1
        <input
          checked={true}
          id="test_radio_op1"
          name="test_radio"
          onChange={[Function]}
          type="radio"
        />
        <span
          className="checkmark"
        />
      </label>
      <label
        className="radio-option"
        htmlFor="test_radio_op2"
      >
        Op2
        <input
          checked={false}
          id="test_radio_op2"
          name="test_radio"
          onChange={[Function]}
          type="radio"
        />
        <span
          className="checkmark"
        />
      </label>
      <label
        className="radio-option"
        htmlFor="test_radio_op3"
      >
        Op3
        <input
          checked={false}
          id="test_radio_op3"
          name="test_radio"
          onChange={[Function]}
          type="radio"
        />
        <span
          className="checkmark"
        />
      </label>
      <label
        className="radio-option"
        htmlFor="test_radio_op4"
      >
        Op4
        <input
          checked={false}
          id="test_radio_op4"
          name="test_radio"
          onChange={[Function]}
          type="radio"
        />
        <span
          className="checkmark"
        />
      </label>
    </div>
  </label>
</div>
`;

exports[`RadioField renders correctly based on the selected value 2`] = `
<div
  className="form-field"
>
  <label
    htmlFor="test_radio_radio_field"
  >
    <span
      className="label-text"
    >
      Radio Test
    </span>
    <div
      className="radio-options"
    >
      <label
        className="radio-option"
        htmlFor="test_radio_op1"
      >
        Op1
        <input
          checked={false}
          id="test_radio_op1"
          name="test_radio"
          onChange={[Function]}
          type="radio"
        />
        <span
          className="checkmark"
        />
      </label>
      <label
        className="radio-option"
        htmlFor="test_radio_op2"
      >
        Op2
        <input
          checked={false}
          id="test_radio_op2"
          name="test_radio"
          onChange={[Function]}
          type="radio"
        />
        <span
          className="checkmark"
        />
      </label>
      <label
        className="radio-option"
        htmlFor="test_radio_op3"
      >
        Op3
        <input
          checked={true}
          id="test_radio_op3"
          name="test_radio"
          onChange={[Function]}
          type="radio"
        />
        <span
          className="checkmark"
        />
      </label>
      <label
        className="radio-option"
        htmlFor="test_radio_op4"
      >
        Op4
        <input
          checked={false}
          id="test_radio_op4"
          name="test_radio"
          onChange={[Function]}
          type="radio"
        />
        <span
          className="checkmark"
        />
      </label>
    </div>
  </label>
</div>
`;

exports[`TextField correctly renders its initial properties 1`] = `
<div
  className="form-field"
>
  <label
    htmlFor="test_text_text_field"
  >
    <span
      className="label-text"
    >
      Test Text
    </span>
    <input
      id="test_text_text_field"
      name="test_text"
      onChange={[Function]}
      type="text"
      value="Testing text field"
    />
  </label>
</div>
`;
